kyaml/kio/byteio_reader.go

```diff
@@ -245,6 +260,55 @@ func (r *ByteReader) Read() ([]*yaml.RNode, error) {
 	return output, nil
 }
 
+// addSeqIndentAnno adds the sequence indentation annotation to the resource
+// value is the input yaml string and node is the decoded equivalent of value
+// the annotation value is decided by deriving the existing sequence indentation of resource
+func addSeqIndentAnno(value string, node *yaml.RNode) error {
+	anno := node.GetAnnotations()
+	if anno[kioutil.SeqIndentAnnotation] != "" {
+		// the annotation already exists, so don't change it
+		return nil
+	}
+
+	currentDefaultIndent := yaml.SequenceIndentationStyle()
+	defer yaml.SetSequenceIndentationStyle(currentDefaultIndent)
+
+	// encode the node to string with default 2 space sequence indentation and calculate the diff
+	yaml.SetSequenceIndentationStyle(yaml.WideSequenceStyle)
+	n, err := yaml.Parse(value)
+	if err != nil {
+		return err
+	}
+	out, err := n.String()
+	if err != nil {
+		return err
+	}
+	twoSpaceIndentDiff := copyutil.PrettyFileDiff(out, value)
+
+	// encode the node to string with compact 0 space sequence indentation and calculate the diff
+	yaml.SetSequenceIndentationStyle(yaml.CompactSequenceStyle)
+	n, err = yaml.Parse(value)
+	if err != nil {
+		return err
+	}
+	out, err = n.String()
+	if err != nil {
+		return err
+	}
+	noIndentDiff := copyutil.PrettyFileDiff(out, value)
+
+	var style string
+	if len(noIndentDiff) <= len(twoSpaceIndentDiff) {
+		style = yaml.CompactSequenceStyle
+	} else {
+		style = yaml.WideSequenceStyle
+	}
```

Comment:
This function is doing two things:
1. Detecting the seqIndentType (or seqIndentStyle)
2. Setting the annotation.

We should decouple these two into separate functions. That will make 1) testable.

But the big issue I see is that the detection logic depends on setting the package-level SeqIndentStyle (That is assuming the library can't be used in a thread-safe manner). Instead, we can have a utility function at the `yaml` level that provides 1) to us, that will also eliminate the need to set/unset that pkg-level variable.

Happy to discuss offline.

Reply:
Avoided package level configurations. PTAL.

Reply:
I would recommend encapsulating the logic to detect the SeqIndentStyle in a separate function so that it can be independently evolved, benchmarked and tested. I need to think more if the detection logic has any edge cases.